### PR TITLE
[LinkerWrapper] Always pass `-flto` if the linker supports it

### PR DESCRIPTION
**Pull request by jhuber6**
Summary;
Now that we use the linker to do LTO / device linking, we need to inform
the `clang` invocation to use `-flto` so it forwards arguments like
`-On` correctly.

### CODE REVIEW

**Comment by jdenny-ornl** [comment]
Seems like it does what it intends to do.  Thanks for working on it.

However, there's a side effect.  Now that -O1 gets passed along, sometimes it triggers an assert fail for AMD GPU:

```
ld.lld: /tmp/llvm/llvm/lib/Target/AMDGPU/AMDGPUResourceUsageAnalysis.cpp:151: virtual bool llvm::AMDGPUResourceUsageAnalysis::runOnModule(llvm::Module&): Assertion `MF && "function must have been generated already"' failed.
```

On my AMD GPU test system, I see new test fails.  They all use -O1 and fail that assert:

```
Failed Tests (4):
  libomptarget :: amdgcn-amd-amdhsa :: api/omp_dynamic_shared_memory_amdgpu.c
  libomptarget :: amdgcn-amd-amdhsa :: api/omp_dynamic_shared_memory_mixed_amdgpu.c
  libomptarget :: amdgcn-amd-amdhsa :: offloading/bug51781.c
  libomptarget :: amdgcn-amd-amdhsa :: offloading/bug51982.c
```

**Comment by jhuber6** [comment]
I'm very familiar with that error message, it's https://github.com/llvm/llvm-project/issues/64863. I guess I should just manually pass the optimization level at `O2` for now.

**Comment by jdenny-ornl** [comment]
> I guess I should just manually pass the optimization level at `O2` for now.

Would -O3 still pass through?

**Comment by jhuber6** [comment]
> > I guess I should just manually pass the optimization level at `O2` for now.
> 
> Would -O3 still pass through?

Should, the behavior that was apparently passing was using `-O2`.

**Comment by jhuber6** [comment]
Updated to pass `-O2` instead, this was the actual behavior that passed, so it's mostly not a functional change.

**Comment by jhuber6** [comment]
We really need to turn off clang-format for the tests, ignoring it.

**Comment by jdenny-ornl** [comment]
> Updated to pass `-O2` instead, this was the actual behavior that passed, so it's mostly not a functional change.

So you just changed the tests not to use -O1?  Doesn't this patch then represent a regression in what's supported?

(Rewriting history with force pushes make it harder to quickly spot what's changed since I looked before.  Can you not push new commits instead?)

File: offload/test/jit/empty_kernel_lvl2.c [hunk 1]
```diff
@@ -21,7 +21,6 @@
 // RUN: env LIBOMPTARGET_JIT_PRE_OPT_IR_MODULE=%t.pre.ll     \
 // RUN:     LIBOMPTARGET_JIT_SKIP_OPT=true                   \
 // RUN:     %libomptarget-run-generic
-// TODO:
```
Comment (jdenny-ornl) [inline offload/test/jit/empty_kernel_lvl2.c:24]: Unrelated changed?

**Comment by jhuber6** [comment]
> > Updated to pass `-O2` instead, this was the actual behavior that passed, so it's mostly not a functional change.
> 
> So you just changed the tests not to use -O1? Doesn't this patch then represent a regression in what's supported?
>

I wouldn't consider something "supported" if it only worked because we overruled the user's `-O1` flag and replaced it with `-O2`.

> (Rewriting history with force pushes make it harder to quickly spot what's changed since I looked before. Can you not push new commits instead?)

Dealing with multiple commits is a pain, I miss Phabricator's diff based review system. Only the tests changed.

File: offload/test/jit/empty_kernel_lvl2.c (hunk 1 above)
Comment (jhuber6) [reply offload/test/jit/empty_kernel_lvl2.c:24]: Yeah, can remove it if needed.

**Comment by jdenny-ornl** [comment]
> > > Updated to pass `-O2` instead, this was the actual behavior that passed, so it's mostly not a functional change.
> > 
> > 
> > So you just changed the tests not to use -O1? Doesn't this patch then represent a regression in what's supported?
> 
> I wouldn't consider something "supported" if it only worked because we overruled the user's `-O1` flag and replaced it with `-O2`.

Cases that used to compile successfully (even if more optimized than expected) will no longer.  Sounds like a regression.

File: offload/test/jit/empty_kernel_lvl2.c (hunk 1 above)
Comment (jdoerfert) [reply offload/test/jit/empty_kernel_lvl2.c:24]: Leave the TODOs please.

File: offload/test/api/omp_dynamic_shared_memory_amdgpu.c [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: %libomptarget-compile-amdgcn-amd-amdhsa -O1 -mllvm -openmp-opt-inline-device
+// RUN: %libomptarget-compile-amdgcn-amd-amdhsa -O2 -mllvm -openmp-opt-inline-device
```
Comment (jdoerfert) [inline offload/test/api/omp_dynamic_shared_memory_amdgpu.c:1]: Why O2 now?

**Review by jdoerfert — APPROVED** [review]
LG, but keep the TODOs.

**Comment by jhuber6** [comment]
> Cases that used to compile successfully (even if more optimized than expected) will no longer. Sounds like a regression.

It kind of is, but I don't think we should work around backend bugs in the clang driver.

**Comment by ajarmusch** [comment]
This caused a nvlink error with clang. This was found by a OpenMP CI https://gitlab.e4s.io/uo-public/llvm-openmp-offloading/-/jobs/301520

**Comment by jhuber6** [comment]
> This caused a nvlink error with clang. This was found by a OpenMP CI https://gitlab.e4s.io/uo-public/llvm-openmp-offloading/-/jobs/301520

Probably an LTO flag I forgot to handle correctly, can you compile it with `-v` to show what it's passing?

**Pull request merged**